haystack/dataclasses/tool.py

```diff
@@ -0,0 +1,242 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import inspect
+from dataclasses import asdict, dataclass
+from typing import Any, Callable, Dict, Optional
+
+from pydantic import create_model
+
+from haystack.lazy_imports import LazyImport
+from haystack.utils import deserialize_callable, serialize_callable
+
+with LazyImport(message="Run 'pip install jsonschema'") as jsonschema_import:
+    from jsonschema import Draft202012Validator
+    from jsonschema.exceptions import SchemaError
+
+
+class ToolInvocationError(Exception):
+    """
+    Exception raised when a Tool invocation fails.
+    """
+
+    pass
+
+
+class SchemaGenerationError(Exception):
+    """
+    Exception raised when automatic schema generation fails.
+    """
+
+    pass
+
+
+@dataclass
+class Tool:
+    """
+    Data class representing a tool for which Language Models can prepare a call.
+
+    Accurate definitions of the textual attributes such as `name` and `description`
+    are important for the Language Model to correctly prepare the call.
+
+    :param name:
+        Name of the tool.
+    :param description:
+        Description of the tool.
+    :param parameters:
+        A JSON schema defining the parameters expected by the tool.
+    :param function:
+        The function that will be invoked when the tool is called.
+    """
+
+    name: str
+    description: str
+    parameters: Dict[str, Any]
+    function: Callable
+
+    def __post_init__(self):
+        jsonschema_import.check()
```

Comment:
Knowing that Tool is central piece of the Agents push, should be we make this dependency default? It's 80Kb binary

Reply:
I'm not totally sure. Let's involve also @julian-risch in the decision.

- I remember that when we introduce new dependencies, we also have to pay attention to Anaconda.
- We had some issues in the past with `jsonschema`: https://github.com/deepset-ai/haystack/pull/7353

Reply:
It seems to be [not occurring](https://github.com/orgs/python-jsonschema/discussions/1237#discussioncomment-8776580) on 3.9 and after, but ok, let's get this integrated and then we can experiment with including it as a default dependency.